### PR TITLE
ARROW-5648: [C++] Avoid using codecvt

### PR DESCRIPTION
**Pull request by pitrou**
Some antiquated C++ build chains miss the standard <codecvt> header.
Use a small vendored UTF8 implementation instead.

### CODE REVIEW

File: cpp/src/arrow/vendored/utf8cpp/core.h [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+        uint32_t ignored;
+        return utf8::internal::validate_next(it, end, ignored);
+    }
+
+} // namespace internal
+
+    /// The library API - functions intended to be called by the users
+
+    // Byte order mark
+    const uint8_t bom[] = {0xef, 0xbb, 0xbf};
```
Comment (fsaintjacques) [inline cpp/src/arrow/vendored/utf8cpp/core.h]: heh, you even have a boom detector now :)

Comment (pitrou) [reply cpp/src/arrow/vendored/utf8cpp/core.h]: boom? :-)

Comment (pitrou) [reply cpp/src/arrow/vendored/utf8cpp/core.h]: We already had one by the way: `SkipUTF8BOM`.

**Comment by nealrichardson** [comment]
Building with this change on the Rtools setup does succeed: https://ci.appveyor.com/project/nealrichardson/arrow/builds/25397807

So as far as that is concerned, LGTM.

**Review by wesm — APPROVED** [review]
+1

**Comment by wesm** [comment]
Appveyor build: https://ci.appveyor.com/project/pitrou/arrow/builds/25396740

**Pull request closed without merging**